src/base/http/connection.cpp

```diff
@@ -33,29 +34,121 @@
 #include <QTcpSocket>
 
 #include "base/logger.h"
-#include "irequesthandler.h"
 #include "requestparser.h"
-#include "responsegenerator.h"
 
 using namespace Http;
 
-Connection::Connection(QTcpSocket *socket, IRequestHandler *requestHandler, QObject *parent)
-    : QObject(parent)
-    , m_socket(socket)
-    , m_requestHandler(requestHandler)
+Connection::Connection(QTcpSocket *socket, QObject *parent)
+    : QObject {parent}
+    , m_socket {socket}
 {
     m_socket->setParent(this);
-    m_idleTimer.start();
-    connect(m_socket, &QTcpSocket::readyRead, this, &Connection::read);
+
+    connect(m_socket, &QAbstractSocket::readyRead, this, &Connection::read);
+    connect(m_socket, &QAbstractSocket::bytesWritten, this, &Connection::bytesWritten);
+    connect(m_socket, &QAbstractSocket::disconnected, this, &Connection::disconnected);
+}
+
+void Connection::sendStatus(const Http::ResponseStatus &status)
+{
+    Q_ASSERT(m_status == PendingStatus);
+    m_socket->write(QString("HTTP/%1 %2 %3")
+                    .arg("1.1", // TODO: depends on request
+                         QString::number(status.code), status.text)
+                    .toLatin1()
+                    .append(Http::CRLF));
+    m_status = PendingHeaders;
+}
+
+void Connection::sendHeaders(const Http::HeaderMap &headers)
+{
+    Q_ASSERT(m_status == PendingHeaders);
+
+    const auto writeHeader = [this](const QString &key, const QString &value)
+    {
+        m_socket->write(QString::fromLatin1("%1: %2")
```

Comment:
can you do a benchmark? compare between before/after this PR that the difference of time spent in preparing a **full** response.
I suspect small writes to the socket drags the performance down.

Reply:
The new code seems a bit slow, I guess have to implement caching in the Connection class

<details>
<summary>new code profile</summary>
<pre>
5266026 function calls (5256250 primitive calls) in 15.027 seconds    

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)        
      3/1    0.000    0.000   15.040   15.040 {built-in method builtins.exec}  
        1    0.000    0.000   15.040   15.040 <string>:1(<module>)
        1    0.018    0.018   15.040   15.040 bench.py:16(run_qbittorrent_test)
      180    0.035    0.000   15.023    0.083 bench.py:4(qbittorrent_test)     
     1260    0.024    0.000   14.178    0.011 sessions.py:463(request)
     1080    0.006    0.000   13.171    0.012 sessions.py:534(get)
      900    0.022    0.000   12.818    0.014 client.py:43(_get)
      900    0.013    0.000   12.793    0.014 client.py:66(_request)
     1260    0.034    0.000   11.694    0.009 sessions.py:614(send)
     1260    0.028    0.000   11.106    0.009 adapters.py:394(send)
      540    0.003    0.000   10.540    0.020 client.py:173(torrents)
     1260    0.041    0.000   10.397    0.008 connectionpool.py:499(urlopen)
     1260    0.034    0.000    9.919    0.008 connectionpool.py:356(_make_request)
     1260    0.011    0.000    9.098    0.007 client.py:1303(getresponse)
     1260    0.025    0.000    9.032    0.007 client.py:300(begin)
     1260    0.023    0.000    8.577    0.007 client.py:267(_read_status)
    15120    0.022    0.000    8.556    0.001 {method 'readline' of '_io.BufferedReader' objects}
     1980    0.010    0.000    8.554    0.004 socket.py:655(readinto)
     1980    8.541    0.004    8.541    0.004 {method 'recv_into' of '_socket.socket' objects}
     1260    0.014    0.000    1.325    0.001 sessions.py:689(merge_environment_settings)
      180    0.001    0.000    1.313    0.007 client.py:158(get_log)
     1260    0.004    0.000    1.279    0.001 utils.py:760(get_environ_proxies)
     1260    0.022    0.000    1.121    0.001 sessions.py:423(prepare_request)
      180    0.002    0.000    1.093    0.006 client.py:12(__init__)
</pre>
</details>

<details>
<summary>oldcode profile</summary>
<pre>
5243886 function calls (5234110 primitive calls) in 13.553 seconds    

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)    
    3/1    0.000    0.000   13.566   13.566 {built-in method builtins.exec}  
        1    0.000    0.000   13.566   13.566 <string>:1(<module>)
        1    0.034    0.034   13.566   13.566 bench.py:16(run_qbittorrent_test)
      180    0.032    0.000   13.521    0.075 bench.py:4(qbittorrent_test)     
     1260    0.023    0.000   12.722    0.010 sessions.py:463(request)
     1080    0.006    0.000   11.863    0.011 sessions.py:534(get)
      900    0.019    0.000   11.454    0.013 client.py:43(_get)
      900    0.013    0.000   11.432    0.013 client.py:66(_request)
     1260    0.031    0.000   10.447    0.008 sessions.py:614(send)
     1260    0.025    0.000    9.945    0.008 adapters.py:394(send)
      540    0.002    0.000    9.617    0.018 client.py:173(torrents)
     1260    0.037    0.000    9.298    0.007 connectionpool.py:499(urlopen)
     1260    0.030    0.000    8.978    0.007 connectionpool.py:356(_make_request)
     1260    0.009    0.000    8.521    0.007 client.py:1303(getresponse)
     1260    0.023    0.000    8.462    0.007 client.py:300(begin)
     1260    0.021    0.000    8.046    0.006 client.py:267(_read_status)
    15300    0.020    0.000    8.027    0.001 {method 'readline' of '_io.BufferedReader' objects}
     1800    0.009    0.000    8.016    0.004 socket.py:655(readinto)
     1800    8.003    0.004    8.003    0.004 {method 'recv_into' of '_socket.socket' objects}
     1260    0.013    0.000    1.190    0.001 sessions.py:689(merge_environment_settings)
     1260    0.004    0.000    1.147    0.001 utils.py:760(get_environ_proxies)
      180    0.001    0.000    1.139    0.006 client.py:158(get_log)
      180    0.002    0.000    1.115    0.006 client.py:12(__init__)
     1260    0.021    0.000    1.049    0.001 sessions.py:423(prepare_request)
</pre>
</details>

<details>
<summary>profile code</summary>
<pre>
from qbittorrent import Client
import cProfile

def qbittorrent_test():
    qb = Client('http://127.0.0.1:8080/')
    qb.login('admin', 'admin123')
    # not required when 'Bypass from localho
    for i in range(3):
        torrents = qb.torrents()

    qb.get_log()
    qb.logout()

def run_qbittorrent_test(n):
    for i in range(n):
        qbittorrent_test()


pr = cProfile.Profile()
pr.run("run_qbittorrent_test(180)")
pr.print_stats("cumulative")
</pre>
</details>